certora/specs/core/RewardsCoordinator.spec

```diff
@@ -0,0 +1,113 @@
+
+	assert (
+        lastReverted,
+        "First claim passing with token claims means second claim should revert"
+    );
+}
+
+/// status: pass
+rule cumulativeClaimedStrictlyIncreasing(env e, address claimToken, address earner) {
+    uint256 cumulativeClaimedBefore = cumulativeClaimed(earner, claimToken);
+    calldataarg args;
+    method f;
+    f(e, args);
+    uint256 cumulativeClaimedAfter = cumulativeClaimed(earner, claimToken);
+
+    assert(cumulativeClaimedAfter >= cumulativeClaimedBefore);
+}
+
+
+/// Check that transfer amount in RewardsCoordinator.prcocessClaim() is equal to leaf.cumulativeEarnings - cumulativeClaimed
+/// cumulativeEarnings - cumulativeClaimed == balanceAfter - balanceBefore
+/// Initially failed from havoc from token safeTransfer in the calldata. Fixed by having unresolved catch-all calls defined
+/// status: pass
+rule transferredTokensFromClaim(env e, IRewardsCoordinator.RewardsMerkleClaim claim, address recipient) {
+    require claim.tokenLeaves.length == 2;
+    require token1 == claim.tokenLeaves[0].token;
+    require token2 == claim.tokenLeaves[1].token;
+    // if recipient is the RewardsCoordinator, rule breaks since balanceAfter == balanceBefore 
+    require recipient != currentContract;
+
+    address earner = claim.earnerLeaf.earner;
+    uint256 token1BalanceBefore = token1.balanceOf(e, recipient);
+    uint256 token2BalanceBefore = token2.balanceOf(e, recipient);
+    uint256 token1CumulativeClaimed = cumulativeClaimed(earner, claim.tokenLeaves[0].token);
+    uint256 token2CumulativeClaimed = cumulativeClaimed(earner, claim.tokenLeaves[1].token);
+
+    processClaim(e, claim, recipient);
+
+    uint256 token1BalanceAfter = token1.balanceOf(e, recipient);
+    uint256 token2BalanceAfter = token2.balanceOf(e, recipient);
+
+    if (token1 == token2) {
+        // In practice this shouldn't occur as we will not construct leaves with multiple tokenLeaves with the same token address
+        assert claim.tokenLeaves[1].cumulativeEarnings - token1CumulativeClaimed == token2BalanceAfter - token2BalanceBefore;
+    } else {
+        assert claim.tokenLeaves[0].cumulativeEarnings - token1CumulativeClaimed == token1BalanceAfter - token1BalanceBefore;
+        assert claim.tokenLeaves[1].cumulativeEarnings - token2CumulativeClaimed == token2BalanceAfter - token2BalanceBefore;
+    }
+}
+
+/// RewardsCoordinator.checkClaim == True => processClaim success 
+/// status: pass
+rule claimCorrectness(env e, IRewardsCoordinator.RewardsMerkleClaim claim, address recipient) {
+    // Call checkClaim to ensure it does not revert in rule since it never returns false
+    checkClaim(e, claim);
+
+    bool canWork = claimerFor(e, claim.earnerLeaf.earner) != 0 ? 
+        e.msg.sender == claimerFor(e, claim.earnerLeaf.earner) :
+        e.msg.sender == claim.earnerLeaf.earner;
+
+    processClaim@withrevert(e, claim, recipient);
+
+    assert !lastReverted => canWork;
```

Comment:
can you assert equality here? that would be stronger if possible; I'm not sure it would work with the summarization of `verifyInclusionKeccak` though...

Reply:
The summarization may be an issue but all other revert reasons will also be problematic. First its things like the contract being paused, in reentrant mode, and >0 msg.value being sent, then compiler checks which are harder to understand.